File: Dock/Globals/Model/LogEntry.js
const crypto = require('crypto');

const { logLevel } = require('../Enumerations/LogLevel');
const { logCategory } = require('../Enumerations/LogCategory');
const { logServiceOrigin } = require('../Enumerations/LogServiceOrigin');

class LogEntry
{
    #id;
    #level;
    #category;
    #title;
    #message;
    #service;
    #accountId;
    #errorCode;
    #errorReason;
    #additionalData;
    #timestamp;
    #timestampIsoString;
    #sequence;
    #environment;

    constructor({level = 1, category = 0, title = '', message = '', service = 0, accountId = '', errorCode = '', errorReason = '', additionalData = {}, timestamp = new Date(), timestampIsoString = '', sequence = 0, environment = ''} = {})
    {
        this.#id = crypto.randomUUID();
        this.setLevel(level);
        this.setCategory(category);
        this.setTitle(title);
        this.setMessage(message);
        this.setService(service);
        this.setAccountId(accountId);
        this.setErrorCode(errorCode);
        this.setErrorReason(errorReason);
        this.setAdditionalData(additionalData);
        this.setTimestamp(timestamp);
        this.setTimestampIsoString(timestampIsoString);
        this.setSequence(sequence);
        this.setEnvironment(environment);
    }

    getId()
    {
        return this.#id;
    }

    getLevel()
    {
        return this.#level;
    }

    setLevel(value)
    {
        if (value !== null)
        {
            const enumValues = Object.values(logLevel);
            if (!enumValues.includes(value))
            {
                value = enumValues[0] ?? null;
            }
        }
        this.#level = value;
    }

    getCategory()
    {
        return this.#category;
    }

    setCategory(value)
    {
        if (value !== null)
        {
            const enumValues = Object.values(logCategory);
            if (!enumValues.includes(value))
            {
                value = enumValues[0] ?? null;
            }
        }
        this.#category = value;
    }

    getTitle()
    {
        return this.#title;
    }

    setTitle(value)
    {
        if (value !== null)
        {
            value = String(value);
        }
        this.#title = value;
    }

    getMessage()
    {
        return this.#message;
    }

    setMessage(value)
    {
        if (value !== null)
        {
            value = String(value);
        }
        this.#message = value;
    }

    getService()
    {
        return this.#service;
    }

    setService(value)
    {
        if (value !== null)
        {
            const enumValues = Object.values(logServiceOrigin);
            if (!enumValues.includes(value))
            {
                value = enumValues[0] ?? null;
            }
        }
        this.#service = value;
    }

    getAccountId()
    {
        return this.#accountId;
    }

    setAccountId(value)
    {
        if (value !== null)
        {
            value = String(value);
        }
        this.#accountId = value;
    }

    getErrorCode()
    {
        return this.#errorCode;
    }

    setErrorCode(value)
    {
        if (value !== null)
        {
            value = String(value);
        }
        this.#errorCode = value;
    }

    getErrorReason()
    {
        return this.#errorReason;
    }

    setErrorReason(value)
    {
        if (value !== null)
        {
            value = String(value);
        }
        this.#errorReason = value;
    }

    getAdditionalData()
    {
        return this.#additionalData;
    }

    setAdditionalData(value)
    {
        this.#additionalData = value;
    }

    getTimestamp()
    {
        return this.#timestamp;
    }

    setTimestamp(value)
    {
        if (value !== null && value !== undefined)
        {
            value = value instanceof Date ? value : new Date(value);
            if (isNaN(value.getTime()))
            {
                value = new Date();
            }
        }
        else
        {
            value = new Date();
        }
        this.#timestamp = value;
    }

    getTimestampIsoString()
    {
        return this.#timestampIsoString;
    }

    setTimestampIsoString(value)
    {
        if (value !== null)
        {
            value = String(value);
        }
        this.#timestampIsoString = value;
    }

    getSequence()
    {
        return this.#sequence;
    }

    setSequence(value)
    {
        if (value !== null)
        {
            value = parseInt(value, 10);
            if (isNaN(value))
            {
                value = 0;
            }
            else
            {
                value = Math.max(value, 0);
            }
        }
        this.#sequence = value;
    }

    getEnvironment()
    {
        return this.#environment;
    }

    setEnvironment(value)
    {
        if (value !== null)
        {
            value = String(value);
        }
        this.#environment = value;
    }

    toJson()
    {
        return {
            id: this.getId(),
            level: this.getLevel() !== null ? Number(this.getLevel()) : null,
            category: this.getCategory() !== null ? Number(this.getCategory()) : null,
            title: this.getTitle(),
            message: this.getMessage(),
            service: this.getService() !== null ? Number(this.getService()) : null,
            accountId: this.getAccountId(),
            errorCode: this.getErrorCode(),
            errorReason: this.getErrorReason(),
            additionalData: this.getAdditionalData(),
            timestamp: this.getTimestamp() !== null ? this.getTimestamp().toISOString() : null,
            timestampIsoString: this.getTimestampIsoString(),
            sequence: this.getSequence(),
            environment: this.getEnvironment(),
        };
    }

    static fromJson(json)
    {
        const instance = new LogEntry({
            level: json.level ?? null,
            category: json.category ?? null,
            title: json.title ?? null,
            message: json.message ?? null,
            service: json.service ?? null,
            accountId: json.accountId ?? null,
            errorCode: json.errorCode ?? null,
            errorReason: json.errorReason ?? null,
            additionalData: json.additionalData ?? null,
            timestamp: json.timestamp != null ? new Date(json.timestamp) : null,
            timestampIsoString: json.timestampIsoString ?? null,
            sequence: json.sequence ?? null,
            environment: json.environment ?? null
        });
        instance._restoreId_id(json.id);
        return instance;
    }

    _restoreId_id(storedId)
    {
        if (storedId !== undefined && storedId !== null)
        {
            this.#id = storedId;
        }
    }
}

module.exports = LogEntry;

File: Agent/Globals/Model/LogEntry.py
import uuid
from datetime import datetime
from Globals.Enumerations.LogLevel import LogLevel
from Globals.Enumerations.LogCategory import LogCategory
from Globals.Enumerations.LogServiceOrigin import LogServiceOrigin


class LogEntry:
    def __init__(self, level: LogLevel = LogLevel(1), category: LogCategory = LogCategory(0), title: str = '', message: str = '', service: LogServiceOrigin = LogServiceOrigin(0), account_id: str = '', error_code: str = '', error_reason: str = '', additional_data: dict = {}, timestamp: datetime = datetime.now(), timestamp_iso_string: str = '', sequence: int = 0, environment: str = '') -> None:
        self.__id = str(uuid.uuid4())
        self.set_level(level)
        self.set_category(category)
        self.set_title(title)
        self.set_message(message)
        self.set_service(service)
        self.set_account_id(account_id)
        self.set_error_code(error_code)
        self.set_error_reason(error_reason)
        self.set_additional_data(additional_data)
        self.set_timestamp(timestamp)
        self.set_timestamp_iso_string(timestamp_iso_string)
        self.set_sequence(sequence)
        self.set_environment(environment)

    def get_id(self) -> str:
        return self.__id

    def get_level(self) -> LogLevel:
        return self.__level

    def set_level(self, value: LogLevel) -> None:
        if value is not None:
            valid_values = list(LogLevel)
            if value not in valid_values:
                value = valid_values[0] if valid_values else None
        self.__level = value

    def get_category(self) -> LogCategory:
        return self.__category

    def set_category(self, value: LogCategory) -> None:
        if value is not None:
            valid_values = list(LogCategory)
            if value not in valid_values:
                value = valid_values[0] if valid_values else None
        self.__category = value

    def get_title(self) -> str:
        return self.__title

    def set_title(self, value: str) -> None:
        if value is not None:
            value = str(value)
        self.__title = value

    def get_message(self) -> str:
        return self.__message

    def set_message(self, value: str) -> None:
        if value is not None:
            value = str(value)
        self.__message = value

    def get_service(self) -> LogServiceOrigin:
        return self.__service

    def set_service(self, value: LogServiceOrigin) -> None:
        if value is not None:
            valid_values = list(LogServiceOrigin)
            if value not in valid_values:
                value = valid_values[0] if valid_values else None
        self.__service = value

    def get_account_id(self) -> str:
        return self.__account_id

    def set_account_id(self, value: str) -> None:
        if value is not None:
            value = str(value)
        self.__account_id = value

    def get_error_code(self) -> str:
        return self.__error_code

    def set_error_code(self, value: str) -> None:
        if value is not None:
            value = str(value)
        self.__error_code = value

    def get_error_reason(self) -> str:
        return self.__error_reason

    def set_error_reason(self, value: str) -> None:
        if value is not None:
            value = str(value)
        self.__error_reason = value

    def get_additional_data(self) -> dict:
        return self.__additional_data

    def set_additional_data(self, value: dict) -> None:
        self.__additional_data = value

    def get_timestamp(self) -> datetime:
        return self.__timestamp

    def set_timestamp(self, value: datetime) -> None:
        if value is not None:
            if isinstance(value, str):
                try:
                    value = datetime.fromisoformat(value)
                except ValueError:
                    value = datetime.now()
            elif not isinstance(value, datetime):
                value = datetime.now()
        else:
            value = datetime.now()
        self.__timestamp = value

    def get_timestamp_iso_string(self) -> str:
        return self.__timestamp_iso_string

    def set_timestamp_iso_string(self, value: str) -> None:
        if value is not None:
            value = str(value)
        self.__timestamp_iso_string = value

    def get_sequence(self) -> int:
        return self.__sequence

    def set_sequence(self, value: int) -> None:
        if value is not None:
            try:
                value = int(value)
                value = max(0, value)
            except (ValueError, TypeError):
                value = 0
        self.__sequence = value

    def get_environment(self) -> str:
        return self.__environment

    def set_environment(self, value: str) -> None:
        if value is not None:
            value = str(value)
        self.__environment = value

    def _restore_id_id(self, stored_id):
        if stored_id is not None:
            self.__id = stored_id

    def to_json(self) -> dict:
        return {
            'id': self.get_id(),
            'level': int(self.get_level().value) if self.get_level() is not None else None,
            'category': int(self.get_category().value) if self.get_category() is not None else None,
            'title': self.get_title(),
            'message': self.get_message(),
            'service': int(self.get_service().value) if self.get_service() is not None else None,
            'accountId': self.get_account_id(),
            'errorCode': self.get_error_code(),
            'errorReason': self.get_error_reason(),
            'additionalData': self.get_additional_data(),
            'timestamp': self.get_timestamp().isoformat() if self.get_timestamp() is not None else None,
            'timestampIsoString': self.get_timestamp_iso_string(),
            'sequence': self.get_sequence(),
            'environment': self.get_environment(),
        }

    @classmethod
    def from_json(cls, data: dict) -> 'LogEntry':
        instance = cls(
            level=LogLevel(data.get('level')) if data.get('level') is not None else None,
            category=LogCategory(data.get('category')) if data.get('category') is not None else None,
            title=data.get('title'),
            message=data.get('message'),
            service=LogServiceOrigin(data.get('service')) if data.get('service') is not None else None,
            account_id=data.get('accountId'),
            error_code=data.get('errorCode'),
            error_reason=data.get('errorReason'),
            additional_data=data.get('additionalData'),
            timestamp=datetime.fromisoformat(data.get('timestamp')) if data.get('timestamp') is not None else None,
            timestamp_iso_string=data.get('timestampIsoString'),
            sequence=data.get('sequence'),
            environment=data.get('environment')
        )
        if data.get('id') is not None:
            instance._restore_id_id(data.get('id'))
        return instance
